### ISSUE TITLE
Cannot use Tensorflow model with batch normalization

### ISSUE THREAD

**Issue by nikolaitasev** [opencv/opencv#15367]
##### System information (version)

- OpenCV => 3.4.2,  3.4.5 and 4.1.0
- Operating System / Platform => Windows and Linux x64
- Compiler => gcc and python bindings

##### Detailed description

There is a large difference in calculations after the first batch normalization layer when I use a convolutional model with the opencv dnn module. The model is made with Keras/Tensorflow 1.14. The graph is frozen and then saved in binary .pb format. 

##### Steps to reproduce

Here is the model up to the batch norm layer in text format
https://www.dropbox.com/s/tujdf3uvd7heyn1/test_batchnorm.pbtxt?dl=0
Here is the model up to the batch norm layer in binary format
https://www.dropbox.com/s/eyndn1xh42x3qr3/test_batchnorm.pb?dl=0

Here is a test image ( grayscale, should be rescaled in range 0..1, before passing as input)
https://www.dropbox.com/s/gtghtp06k8jk5vy/00001.png?dl=0
And the outputs generated with the python binding
Here is the output of the batch norm in Tensorflow
https://www.dropbox.com/sh/xwkdjwbsoarxqna/AADrZ9BmWK5UUGRBN9t1I6moa?dl=0
Here is the output of the batch norm in OpenCV
https://www.dropbox.com/sh/520m2w74045tlqc/AAAUl3CAzE0r498go7BsMxRea?dl=0

**Comment by dkurt** [comment]
@nikolaitasev, Hi!

A bit confused about nodes:
```
name: "conv2d/Conv2D"
op: "Conv2D"
input: "conv2d_input"
input: "conv2d/Conv2D/ReadVariableOp"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "data_format"
  value {
    s: "NHWC"
  }
}
attr {
  key: "dilations"
  value {
    list {
      i: 1
      i: 1
      i: 1
      i: 1
    }
  }
}
attr {
  key: "explicit_paddings"
  value {
    list {
    }
  }
}
attr {
  key: "padding"
  value {
    s: "VALID"
  }
}
attr {
  key: "strides"
  value {
    list {
      i: 1
      i: 1
      i: 1
      i: 1
    }
  }
}
attr {
  key: "use_cudnn_on_gpu"
  value {
    b: true
  }
}
```

and

```
name: "conv2d/BiasAdd"
op: "BiasAdd"
input: "conv2d/Conv2D"
input: "conv2d/BiasAdd/ReadVariableOp"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "data_format"
  value {
    s: "NHWC"
  }
}
```

but

```
name: "batch_normalization/FusedBatchNorm"
op: "FusedBatchNorm"
input: "conv2d/Relu"
input: "batch_normalization/ReadVariableOp"
input: "batch_normalization/ReadVariableOp_1"
input: "batch_normalization/FusedBatchNorm/ReadVariableOp"
input: "batch_normalization/FusedBatchNorm/ReadVariableOp_1"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "data_format"
  value {
    s: "NCHW"
  }
}
attr {
  key: "epsilon"
  value {
    f: 0.0010000000474974513
  }
}
attr {
  key: "is_training"
  value {
    b: false
  }
}
```

So both convolution and it's bias are NHWC but BatchNorm is NCHW. Can you please share the code which generate this network? Thanks in advance!

**Comment by nikolaitasev** [comment]
I found the problem when you pointed the data_format mismatch.
For the training I used an old code that was in channel first order. 
When I converted to channels last seems I forgot to change the BatchNorm layer.

model = tf.keras.models.Sequential([
   tf.keras.layers.Conv2D(32, (3,3), activation='relu', input_shape=(input_side,input_side, 1)),
   tf.keras.layers.BatchNormalization(axis=1)
  ....
instead of 
  tf.keras.layers.BatchNormalization(axis=-1)

After that I tried again and the sample image was classified correctly.

It is strange why in TensorFlow it works ok even with the wrong code and in OpenCV it doesn't.

**Comment by dkurt** [comment]
@nikolaitasev, Most likely your model were trained to normalize different dimension but not channels. So tensor data was in NHWC order but BatchNorm used it as NCHW so H dimension was really normalized.

**Comment by nikolaitasev** [comment]
@dkurt Yes that I understand. What I don't understand is what is the problem in OpenCV:
  - it doesn't support normalization on the second dimension in BatchNorm
or
  - it requires that the data_format is the same for all layers

**Comment by dkurt** [comment]
@nikolaitasev, OpenCV works in NCHW and performs BatchNorm by channels dimensions only (1st dimension):

|  | Convolution | BatchNorm |
|---|---|---|
| TensorFlow | NHWC | NCHW |
| OpenCV | NCHW | NCHW |

So to perform normalization over H dimension as TensorFlow does in that case, OpenCV should transpose tensor from NCHW to NHWC first then back to NCHW. It's quite inconvenient scenario so we have no layout check before BatchNorm layer.

**Comment by nikolaitasev** [comment]
@dkurt Thanks for the clarification. 
Does this means that it is better if I change the saved model so that all layers are NCHW?
Will it bring me more performance than the default layout of Tensorflow NHWC?

**Comment by dkurt** [comment]
@nikolaitasev, it's better to keep the default layout, NHWC. According to TensorFlow [docs](https://www.tensorflow.org/guide/performance/overview), these days both implementations are well optimized.

**Comment by nikolaitasev** [comment]
@dkurt Thanks for the reply.
